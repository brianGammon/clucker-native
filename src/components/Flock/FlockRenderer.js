import React from 'react';
import {
  Text,
  Button,
  View,
  FlatList,
  TouchableOpacity,
  Image,
} from 'react-native';
import { type Chicken, type Navigation, type Flock } from '../../types';

type Props = {
  navigation: Navigation,
  chickens: {
    [string]: Chicken,
  },
  flock: Flock,
  isFlockOwner: boolean,
};

const FlockRenderer = ({
  navigation,
  chickens,
  flock,
  isFlockOwner,
}: Props) => (
  <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
    <Text>Flock Name: {flock && flock.name}</Text>
    <FlatList
      style={{ width: '100%' }}
      data={Object.keys(chickens || {})}
      keyExtractor={(item, index) => index.toString()}
      renderItem={({ item }) => (
        <TouchableOpacity
          style={{
            flex: 1,
            flexDirection: 'row',
            borderBottomWidth: 1,
            borderBottomColor: 'grey',
            padding: 10,
          }}
          onPress={() => navigation.navigate('Chicken', { chickenId: item })}
        >
          <View>
            <Image
              style={{ width: 80, height: 80 }}
              source={
                chickens[item].thumbnailUrl
                  ? { uri: chickens[item].thumbnailUrl }
                  : require('../../assets/default-profile-photo_thumb.png')
              }
            />
          </View>
          <View style={{ flex: 1, paddingLeft: 10 }}>
            <Text>Name: {chickens[item].name}</Text>
          </View>
        </TouchableOpacity>
      )}
    />
    {isFlockOwner && (
      <Button
        onPress={() => navigation.navigate('ChickenEditor')}
        title="Add a Chicken"
      />
    )}
  </View>
);

export default FlockRenderer;
