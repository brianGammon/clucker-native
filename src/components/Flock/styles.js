import { StyleSheet } from 'react-native';
import variables from 'native-base/dist/src/theme/variables/platform';

export default StyleSheet.create({
  noChickens: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  noChickensText: {
    color: variables.inputColorPlaceholder,
  },
  header: {
    flexDirection: 'row',
    justifyContent: 'center',
  },
  headerBlock: {
    alignItems: 'center',
  },
  headerCell: {
    flex: 1,
    alignItems: 'center',
  },
  headerCellLabel: {
    fontSize: 16,
    fontWeight: 'bold',
    color: variables.inputColorPlaceholder,
  },
  last: {
    borderLeftWidth: 1,
    borderLeftColor: variables.inputColorPlaceholder,
  },
  bodyCell: {
    flex: 1,
  },
  trophyCell: {
    alignItems: 'center',
    marginRight: 14,
  },
  trophyCellText: {
    fontSize: 12,
  },
  trophyIcon: {
    fontSize: 24,
  },
  addText: {
    paddingLeft: 8,
    color: variables.inputColorPlaceholder,
  },
});
