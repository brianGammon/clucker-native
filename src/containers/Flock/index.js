/* @flow */
import * as React from 'react';
import { connect } from 'react-redux';
import {
  Text, Button, View, FlatList, TouchableOpacity, Image,
} from 'react-native';
import { type Chicken, type Navigation } from '../../types';

type Props = {
  navigation: Navigation,
  chickens: {
    [string]: Chicken,
  },
};

class Flock extends React.Component<Props> {
  render() {
    const { navigation, chickens } = this.props;
    return (
      <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
        <FlatList
          style={{ width: '100%' }}
          data={Object.keys(chickens || {})}
          keyExtractor={(item, index) => index.toString()}
          renderItem={({ item }) => (
            <TouchableOpacity
              style={{
                flex: 1,
                flexDirection: 'row',
                borderBottomWidth: 1,
                borderBottomColor: 'grey',
                padding: 10,
              }}
              onPress={() => navigation.navigate('Chicken')}
            >
              <View>
                <Image
                  style={{ width: 80, height: 80 }}
                  source={{ uri: chickens[item].thumbnailUrl }}
                />
              </View>
              <View style={{ flex: 1, paddingLeft: 10 }}>
                <Text>Name: {chickens[item].name}</Text>
                <Text>Breed: {chickens[item].breed}</Text>
                <Text>Hatched On: {chickens[item].hatched}</Text>
              </View>
            </TouchableOpacity>
          )}
        />
        <Button onPress={() => navigation.navigate('ChickenEditor')} title="Add a Chicken" />
        <Button onPress={() => navigation.navigate('Chicken')} title="Chicken Profile" />
      </View>
    );
  }
}

const mapStateToProps = ({ chickens: { items } }) => ({ chickens: items });
export default connect(mapStateToProps)(Flock);
