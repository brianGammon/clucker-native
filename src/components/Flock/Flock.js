/* @flow */
import * as React from 'react';
import { connect } from 'react-redux';
import FlockRenderer from './FlockRenderer';
import {
  type Chicken,
  type Navigation,
  type Flock as FlockType,
} from '../../types';
import { firebaseRemoveRequested } from '../../redux/actions';
import currentFlockSelector from '../../selectors/currentFlockSelector';
import isFlockOwnerSelector from '../../selectors/isFlockOwnerSelector';
import { metaTypes } from '../../redux/constants';

type Props = {
  navigation: Navigation,
  chickens: {
    [string]: Chicken,
  },
  flockId: string,
  flock: FlockType,
  isFlockOwner: boolean,
  deleteChicken: (flockId: string, chickenId: string) => void,
};

class Flock extends React.Component<Props> {
  constructor() {
    super();

    this.onDeleteChicken = this.onDeleteChicken.bind(this);
  }

  onDeleteChicken = (chickenId) => {
    const { flockId, deleteChicken } = this.props;
    deleteChicken(flockId, chickenId);
  };

  render() {
    const {
      navigation, chickens, flock, isFlockOwner,
    } = this.props;
    return (
      <FlockRenderer
        navigation={navigation}
        chickens={chickens}
        flock={flock}
        isFlockOwner={isFlockOwner}
        onDeleteChicken={this.onDeleteChicken}
      />
    );
  }
}

const mapStateToProps = ({ chickens, flocks, userSettings }) => ({
  chickens: chickens.data,
  flockId: userSettings.data.currentFlockId,
  flock: currentFlockSelector(flocks.data, userSettings),
  isFlockOwner: isFlockOwnerSelector(flocks.data, userSettings),
});

const mapDispatchtoProps = dispatch => ({
  deleteChicken: (flockId, chickenId) => dispatch(
    firebaseRemoveRequested({ flockId, chickenId }, metaTypes.chickens),
  ),
});

export default connect(
  mapStateToProps,
  mapDispatchtoProps,
)(Flock);
