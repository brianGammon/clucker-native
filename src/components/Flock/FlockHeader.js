/* @flow */
import React from 'react';
import {
  View, Text, Button, Icon, H2,
} from 'native-base';
import styles from './styles';
import { type Flock, type Chicken, type Navigation } from '../../types';

type Props = {
  navigation: Navigation,
  flock: Flock,
  chickens: {
    [chickenId: string]: Chicken,
  },
  isFlockOwner: boolean,
};

const FlockHeader = ({
  navigation, flock, chickens, isFlockOwner,
}: Props) => (
  <View padder style={styles.header}>
    <View style={styles.headerCell}>
      <Text style={styles.headerCellLabel}>Flock Name</Text>
      <H2>{flock && flock.name}</H2>
    </View>
    <View style={[styles.header, styles.headerCell, styles.last]}>
      <View style={styles.headerBlock}>
        <Text style={styles.headerCellLabel}>Chickens</Text>
        <H2>{Object.keys(chickens || {}).length}</H2>
      </View>
      {isFlockOwner && (
        <View>
          <Button
            transparent
            dark
            onPress={() => navigation.navigate('ChickenEditor')}
          >
            <Icon active name="add-circle" />
            {/* <Text style={styles.addText}>Chicken</Text> */}
          </Button>
        </View>
      )}
    </View>
  </View>
);

export default FlockHeader;
