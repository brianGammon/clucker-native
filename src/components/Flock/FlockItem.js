/* @flow */
import React from 'react';
import {
  View,
  Text,
  ListItem,
  Thumbnail,
  Left,
  Body,
  Right,
  Icon,
} from 'native-base';
import { type Chicken, type Navigation } from '../../types';
import styles from './styles';

type Props = {
  navigation: Navigation,
  item: string,
  chickens: {
    [chickenId: string]: Chicken,
  },
  topProducer: string,
};

const FlockItem = ({
  item, chickens, navigation, topProducer,
}: Props) => (
  <ListItem
    onPress={() => navigation.navigate('Chicken', { chickenId: item })}
    thumbnail
  >
    <Left>
      <Thumbnail
        square
        source={
          chickens[item].thumbnailUrl
            ? { uri: chickens[item].thumbnailUrl }
            : require('../../assets/default-profile-photo_thumb.png')
        }
      />
    </Left>
    <Body>
      <View style={styles.header}>
        <View style={styles.bodyCell}>
          <Text>{chickens[item].name}</Text>
          <Text note numberOfLines={1}>
            {chickens[item].breed}
          </Text>
        </View>
        {topProducer
          && topProducer === item && (
            <View style={[styles.bodyCell, styles.trophyCell]}>
              <Icon
                style={styles.trophyIcon}
                type="FontAwesome"
                name="trophy"
              />
              <Text style={styles.trophyCellText}>Top Producer</Text>
            </View>
        )}
      </View>
    </Body>
    <Right>
      <Icon name="arrow-forward" />
    </Right>
  </ListItem>
);

export default FlockItem;
