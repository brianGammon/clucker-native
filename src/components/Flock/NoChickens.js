/* @flow */
import React from 'react';
import { View, Text } from 'native-base';
import styles from './styles';

const NoChickens = () => (
  <View style={styles.noChickens}>
    <Text>
      No chickens in your flock yet. Add one by tapping the &#34;+&#34; button
      above.
    </Text>
  </View>
);

export default NoChickens;
