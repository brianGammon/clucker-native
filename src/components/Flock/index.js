import Flock from './Flock';

export default Flock;
