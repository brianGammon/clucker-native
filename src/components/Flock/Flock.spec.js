import React from 'react';
import configureStore from 'redux-mock-store';
import Flock from './Flock';
import FlockRenderer from './FlockRenderer';

describe('Flock component:', () => {
  const middlewares = [];
  const mockStore = configureStore(middlewares);
  const sampleData = {
    chickens: {
      data: {
        chicken1: {
          name: 'Bossie',
          breed: 'Maran',
        },
        chicken2: {
          name: 'Bossie',
          breed: 'Maran',
        },
      },
    },
    flocks: {
      data: {
        flock1: {
          name: 'Test Flock',
        },
        flock2: {
          name: 'Test Flock 2',
        },
      },
    },
    userSettings: {
      data: {
        currentFlockId: 'flock1',
      },
    },
  };
  const store = mockStore(sampleData);

  test('Should render connected component', () => {
    const navigation = {};
    const wrapper = shallow(
      <Flock store={store} navigation={navigation} />,
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });

  test('Should render FlockRenderer', () => {
    const navigation = jest.fn();
    const chickens = sampleData.chickens.data;
    const flock = sampleData.flocks.data.flock2;
    const wrapper = shallow(
      <FlockRenderer
        navigation={navigation}
        chickens={chickens}
        flock={flock}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
