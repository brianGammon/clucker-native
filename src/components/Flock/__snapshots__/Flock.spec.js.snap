// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flock component: Should render FlockRenderer as flock owner 1`] = `
<Component
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Component>
    Flock Name: 
    Test Flock 2
  </Component>
  <FlatList
    data={
      Array [
        "chicken1",
        "chicken2",
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    renderItem={[Function]}
    scrollEventThrottle={50}
    style={
      Object {
        "width": "100%",
      }
    }
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
  <Button
    onPress={[Function]}
    title="Add a Chicken"
  />
</Component>
`;

exports[`Flock component: Should render FlockRenderer not as flock owner 1`] = `
<Component
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Component>
    Flock Name: 
    Test Flock 2
  </Component>
  <FlatList
    data={
      Array [
        "chicken1",
        "chicken2",
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    renderItem={[Function]}
    scrollEventThrottle={50}
    style={
      Object {
        "width": "100%",
      }
    }
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</Component>
`;

exports[`Flock component: Should render connected component as flock guest 1`] = `
<FlockRenderer
  chickens={
    Object {
      "chicken1": Object {
        "breed": "Maran",
        "name": "Bossie",
      },
      "chicken2": Object {
        "breed": "Maran",
        "name": "Bossie",
      },
    }
  }
  flock={
    Object {
      "name": "Test Flock",
      "ownedBy": "user1",
    }
  }
  isFlockOwner={false}
  navigation={Object {}}
  onDeleteChicken={[Function]}
/>
`;

exports[`Flock component: Should render connected component as flock owner 1`] = `
<FlockRenderer
  chickens={
    Object {
      "chicken1": Object {
        "breed": "Maran",
        "name": "Bossie",
      },
      "chicken2": Object {
        "breed": "Maran",
        "name": "Bossie",
      },
    }
  }
  flock={
    Object {
      "name": "Test Flock",
      "ownedBy": "user1",
    }
  }
  isFlockOwner={true}
  navigation={Object {}}
  onDeleteChicken={[Function]}
/>
`;
